walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,183 @@
+    // Find and return these time ranges only if there are still time ranges left.
+    combineTimeRanges(eventTimeRanges, optionalEventTimeRanges);
+    ArrayList<TimeRange> availableCombinedTimeRanges = 
+        getAvailableTimeRanges(eventTimeRanges, request.getDuration());
+    
+    if (availableCombinedTimeRanges.isEmpty()) {
+      // Adding optional events got rid of all possible time slots, so we want to 
+      // return the available time ranges from before we combined mandatory/optional.
+      return availableTimeRanges;
+    }
+
+    return availableCombinedTimeRanges;
+  }
+
+  /**
+   * Build a list of events that contain the given attendees.
+   */
+  private ArrayList<Event> getEventsWithAttendees(Collection<Event> events,
+                                                  Collection<String> attendees) {
+    ArrayList<Event> attendedEvents = new ArrayList<Event>();
+    for (Event event : events) {
+      for (String person : event.getAttendees()) {
+        if (attendees.contains(person)) {
+          attendedEvents.add(event);
+          break;
+        }
+      }
+    }
+
+    return attendedEvents;
+  }
+
+  /**
+   * Given a list of events, get the time ranges that the events take up, merging any 
+   * time ranges that overlap.
+   */
+  private ArrayList<TimeRange> getEventTimeRanges(ArrayList<Event> attendedEvents) {
+    // If there are no events, there are no time ranges to get.
+    if (attendedEvents.size() == 0) {
+      return new ArrayList<TimeRange>();
+    }
+
+    ArrayList<TimeRange> eventTimeRanges = new ArrayList<TimeRange>();
+    eventTimeRanges.add(attendedEvents.get(0).getWhen());
+
+    // Iterate through each event, adding merged time ranges to eventTimeRanges. 
+    for (Event event : attendedEvents) {
+      TimeRange currentTimeRange = event.getWhen();
+      TimeRange previousTimeRange = eventTimeRanges.get(eventTimeRanges.size() - 1);
+      
+      // If the event starts after the previous one ends, add its time range to the stack.
+      if (previousTimeRange.end() < currentTimeRange.start()) {
+        eventTimeRanges.add(currentTimeRange);
+      }
+
+      // Otherwise, if the event ends after the previous one, merge their time ranges.
+      else if (previousTimeRange.end() < currentTimeRange.end()) {
+        eventTimeRanges.remove(eventTimeRanges.size() - 1);
+        eventTimeRanges.add(TimeRange.fromStartEnd(previousTimeRange.start(), 
+                                                   currentTimeRange.end(), 
+                                                   false));
+      }
+    }
+
+    return eventTimeRanges;
+  }
+
+  /**
+   * Combine the time ranges of mandatory and optional events, keeping only the optional
+   * events that do not overlap with mandatory events.
+   */
+  private void combineTimeRanges(ArrayList<TimeRange> mandatory, ArrayList<TimeRange> optional) {
+    int initialSize = mandatory.size();
```

Comment:
Nice! 

\**All of the below is just extra info and you don't actually need to make any changes here.\**

Another option is to return a new list that contains both mandatory && non-overlapping optional time ranges.
You won't need to keep track of initialSize anymore and can use `for(mandatoryTimeRange : mandatory)` instead on ln 148.

There's also this cool thing in Java called [streams](https://winterbe.com/posts/2014/07/31/java8-stream-tutorial-examples/), which lets you operate on a whole collection.  

With it you can do cool things like:
```
HashSet<TimeRange> mandatoryRanges = new HashSet(mandatory);
ArrayList<TimeRange> nonConflictingOptionalTimeRanges = optional.stream()
  .filter(optionalTimeRange -> !mandatoryRanges.contains(optionalTimeRange)
  .collect(Collectors.toCollection(ArrayList::new));

ArrayList<TimeRange> mergedEvents = new ArrayList<TimeRange>();
mergedEvents.addAll(mandatory);
mergedEvents.addAll(nonConflictingOptionalTimeRanges);
// sort mergedEvents...
return mergedEvents;
```

Sorry I get excited about Streams haha. They're probably my favorite part about Java :P

Reply:
Awesome, I'll definitely consider using streams in the future. I figured the way I was combining the lists of TimeRanges was kind of awkward, so it's cool to see a more elegant alternative :)